### DISCUSSION TITLE
Dynamic Forms

### DISCUSSION THREAD

**Discussion by tetienne** [charmbracelet/huh#582] [Ideas]
**Is your feature request related to a problem? Please describe.**
My current CLI ask several questions to the end user, and each question depends on the choice of the previous one. For instance, select an AWS cluster, then an AWS service.
So I cannot use the form feature of huh (or can I?). Because of this, I cannot display the help message for my prompt.

**Describe the solution you'd like**
Add a parameter ShowHelpMessage for the fields also.

**Describe alternatives you've considered**
Create a form for each single question.

**Additional context**
N/A

**Comment by meowgorithm** [comment]
Hi! This is something we'd to explore in a more official capacity, but for now you can achieve a similar effect using multiple forms. For an example to help get started see https://github.com/charmbracelet/huh/pull/75.

**Comment by tetienne** [comment]
Thx for the exemple. So no magic yet here :)

**Comment by alinagard** [comment]
+1 to this enhancement!!

**Comment by kpym** [comment]
You can may be use `WithHideFunc` to conditionally show some questions. Here is the rewriten [example](https://github.com/charmbracelet/huh/blob/0f20f61417bfee7a1ec6f0f2f93a423d884b20ad/examples/conditional/main.go) of #75 (with corrected "Carrot" option for "vegetables" 🙂).

```golang
package main

import (
	"fmt"

	"github.com/charmbracelet/huh"
)

func main() {

	var (
		category string
		choice   string
	)

	huh.NewForm(
		huh.NewGroup(
			huh.NewSelect[string]().
				Title("What are you in the mood for?").
				Value(&category).
				Options(
					huh.NewOption("Some fruit", "fruits"),
					huh.NewOption("A vegetable", "vegetables"),
					huh.NewOption("A drink", "drinks"),
				),
		),
		huh.NewGroup(
			huh.NewSelect[string]().
				Title("Okay, what kind of fruits are you in the mood for?").
				Value(&choice).
				Options(
					huh.NewOption("Tangerine", "tangerine"),
					huh.NewOption("Canteloupe", "canteloupe"),
					huh.NewOption("Pomelo", "pomelo"),
					huh.NewOption("Grapefruit", "grapefruit"),
				),
		).WithHideFunc(func() bool {
			return category != "fruits"
		}),
		huh.NewGroup(
			huh.NewSelect[string]().
				Title("Okay, what kind of vegetables are you in the mood for?").
				Value(&choice).
				Options(
					huh.NewOption("Carrot", "carrot"),
					huh.NewOption("Jicama", "jicama"),
					huh.NewOption("Kohlrabi", "kohlrabi"),
					huh.NewOption("Fennel", "fennel"),
					huh.NewOption("Ginger", "ginger"),
				),
		).WithHideFunc(func() bool {
			return category != "vegetables"
		}),
		huh.NewGroup(
			huh.NewSelect[string]().
				Title("Okay, what kind of drinks are you in the mood for?").
				Value(&choice).
				Options(
					huh.NewOption("Coffee", "coffee"),
					huh.NewOption("Tea", "tea"),
					huh.NewOption("Bubble Tea", "bubble-tea"),
					huh.NewOption("Agua Fresca", "agua-fresca"),
				),
		).WithHideFunc(func() bool {
			return category != "drinks"
		}),
	).Run()

	fmt.Printf("One %s coming right up!\n", choice)
}
```

**Comment by iloveicedgreentea** [comment]
I would like to extend this request to include external updates such as API calls. I mentioned some details here https://github.com/charmbracelet/huh/issues/183

**Comment by meowgorithm** [comment]
Yep, that's definitely something we'll need to support, which also means we'll probably need a spinner if the `func` that computes the data runs over a certain time threshold.

**Comment by maaslalani** [comment]
Just a heads up we're making good progress on dynamic forms. It looks like the API will be mostly allowing any option `Title`, `Description`, `Option` to be "Func"-able, i.e. take a function to populate the field from. It will automatically cache the options if desired as well.


https://github.com/charmbracelet/huh/assets/42545625/cdff62f1-a3aa-4a19-93f0-d002b96bbe2d

**Comment by iloveicedgreentea** [comment]
Awesome really looking forward to it. I really enjoy working with Huh, it has been great and everyone has been super impressed of the things we used it for

**Comment by yann-soubeyrand** [comment]
Hello, I’d like to have a group where I can dynamically add/remove input fields with key bindings (like pressing `n` to add a field and pressing `d` to remove the current field). Will the work on this issue cover this use case or is it something completely unrelated?

**Comment by nilansaha** [comment]
I am aware of the example on conditional form but is there any way to populate the options in a group conditionally as well? @maaslalani

**Reply by bashbunni** [reply]
Hey @nilansaha take a look at this example: https://github.com/charmbracelet/huh/blob/main/examples/dynamic/dynamic-country/main.go

It changes the `state` options based on the selected `country`. Please let me know if that's what you're looking for!

**Comment by neilyoung** [comment]
Dynamics work fine, but since `WithHidefunc` only works for groups, there is kind of an interruption in the user experience. I personally would like to have an overview about all my decisions made through the order process, but once a new group appears on screen, the history is lost. Is there any way around, so that all previous decisions are still visible –~and could even be reverted later on~?

Edit: So to say a "WithHideFunc" per "NewSelect" would be great.

**Reply by bashbunni** [reply]
Hey @neilyoung do you have any kind of mock up or example of this behaviour that you could showcase here? I want to make sure I'm understanding your vision. Thank you!!

**Reply by neilyoung** [reply]
@bashbunni No sorry. I dropped it and don’t remember anything

**Reply by bashbunni** [reply]
Hey @neilyoung no worries, sorry it took so long to get back to you 🙏 thanks for the reply :)

**Reply by miniscruff** [reply]
Hi, I think I have a use case for this could I build a POC and show it?

**Reply by bashbunni** [reply]
Hey @miniscruff that would be really helpful! I'll keep an eye on this thread

**Reply by miniscruff** [reply]
ok created a PR https://github.com/charmbracelet/huh/pull/597 which also links to my own projects PR here https://github.com/miniscruff/changie/pull/685 it is very rough and incomplete ( most fields are not dynamic only the ones I needed to test )

**Reply by skatkov** [reply]
It's a great mock! Exactly what I need.

I have form that would require additional questions if some options are picked and this mock would solve that great.

**Reply by miniscruff** [reply]
@bashbunni definitely don't want to be pushy but just wanted to ping you just in case this was lost in all your notifications.

**Comment by bashbunni** [comment]
Hey I believe the original concept of this feature has been implemented as Huh does support dynamic forms. That said, there continues to be a lot of discussion around this feature, so I'm going to convert this to a Discussion so we can continue to track community sentiment towards dynamic forms. Once we decide to go forward with a new feature and are ready to start work on it, we will create a new issue outlining the scope of that feature.

Thank you all so much for wanting to make Huh better. We will be looking at these discussions when looking for new features to implement going forward.

**Comment by zevdg** [comment]
So I was playing around with the dynamic forms feature, and although they're really cool, they don't do the thing I would like to use them for: dynamic values, or at least dynamic value defaults. Is there some way to do this that I'm not seeing?

The use case here is when you have 2 fields in a form that are related such that a good default for one of them can be derived from the other one. Unfortunately, [the current mechanism for setting value defaults](https://github.com/charmbracelet/huh/issues/55#issuecomment-1855911142) appears to be entirely static.  I was hoping to find a `ValueFunc` or `DefaultFunc` similar to `PlaceholderFunc`. 

I also tried to implement this behavior via custom Accessors - first with a setter func that set multiple values when one field was updated, and then separately with a getter func that reads from multiple fields. Neither worked as expected. ~It seems like huh only updates the view for a field when it gets a signal indicating that the field has been edited (either directly, or via the dynamic *Func's bindings args).  I don't see any obvious mechanism for triggering that signal from or for multi-field accessors.~ _edit: On closer inspection, it looks like huh just treats value as a one-way binding where `textinput.Model.Value()` is considered the source of truth and out-of-band updates to the bound value are simply ignored and clobbered._ 

I figured I should post on this discussion before digging deeper to see if I'm missing something obvious.

Also, is this a feature that anyone else would be interested in? If so, I'd be happy to take a stab at it.

**Reply by zevdg** [reply]
I went ahead and made a concrete proposal in PR https://github.com/charmbracelet/huh/pull/626 to keep this conversation moving forward.  That PR adds `Default` and `DefaultFunc` methods to the Input struct.

**Reply by skatkov** [reply]
This is the issue I wanted to raise myself, current implementation of dynamic forms is a great start but feels lacking. And kudos to mister @zevdg he not only highlighted the issue I'm facing, but also proposed a solution.

Hopefully, his PR will be able to move forward!

**Comment by schrodienieur** [comment]
How to create dynamics fields?

I have productSku NewInput, after the user input the productSku, the next fields should be depends on the API call response.

The API response have a structure like this:
```go

type ProductsVariantsResponse struct {
	CatalogAPI
	Data *[]ProductsVariantsResponseData `json:"data"`
}

 type ProductsVariantsResponseData struct {
	AttributeLabel       string   `json:"attribute_label"`
	AttributeName        string   `json:"attribute_name"`
	AttributeOrder       int      `json:"attribute_order"`
	IsSwatch             bool     `json:"is_swatch"`
	AttributeDescription string   `json:"attribute_description"`
	Options              []Option `json:"options"`
}

type Option struct {
	Label    string `json:"option_label"`
	Value    string `json:"option_value"`
	Position int    `json:"option_position"`
}
```

The response is nested, So there will be many attribute with many options.

Let say, i got 3 attributes with 3 options in every attribute. 

- AttributeLabel 1
  - Option.Label 1
  - Option.Label 2
  - Option.Label 3
- AttributeLabel 2
  - Option.Label 1
  - Option.Label 2
  - Option.Label 3
- AttributeLabel 3
  - Option.Label 1
  - Option.Label 2
  - Option.Label 3

How to make 3 NewSelect with 3 NewOption dynamically? I think the current implementation only supports dynamic options.
